pageserver/src/tenant/timeline.rs

```diff
@@ -2972,7 +3079,13 @@ impl Timeline {
         let _write_guard = if write_lock_held {
             None
         } else {
-            Some(self.write_lock.lock().await)
+            let mut writer_guard = self.write_lock.lock().await;
+
+            // Drop any writer state that refers to the layer we're about to freeze: writers may
+            // only hold references to an unfrozen layer, or they'll assert out when they try to use it.
+            writer_guard.take();
```

Comment:
Not a big fan of dropping the state here. It's hard to see what's going on in the `TimelineWriter` if you're not aware of the drop squirrelled away here. I'd rather encapsulate a hack in `TimelineWriter`, but if you want to keep it as is, let's add a comment somewhere in the writer that points to this.